packages/cli/src/ui/components/shared/text-buffer.ts

```diff
@@ -395,187 +395,103 @@ function calculateVisualLayout(
-  useEffect(() => {
-    const layout = calculateVisualLayout(
-      lines,
-      [cursorRow, cursorCol],
-      viewport.width,
-    );
-    setVisualLines(layout.visualLines);
-    setVisualCursor(layout.visualCursor);
-    setLogicalToVisualMap(layout.logicalToVisualMap);
-    setVisualToLogicalMap(layout.visualToLogicalMap);
-  }, [lines, cursorRow, cursorCol, viewport.width]);
-
-  // Update visual scroll (vertical)
-  useEffect(() => {
-    const { height } = viewport;
-    let newVisualScrollRow = visualScrollRow;
+interface TextBufferState {
+  lines: string[];
+  cursorRow: number;
+  cursorCol: number;
+  preferredCol: number | null; // This is visual preferred col
+  undoStack: UndoHistoryEntry[];
+  redoStack: UndoHistoryEntry[];
+  clipboard: string | null;
+  selectionAnchor: [number, number] | null;
+}
 
-    if (visualCursor[0] < visualScrollRow) {
-      newVisualScrollRow = visualCursor[0];
-    } else if (visualCursor[0] >= visualScrollRow + height) {
-      newVisualScrollRow = visualCursor[0] - height + 1;
+const historyLimit = 100;
+
+type TextBufferAction =
+  | { type: 'SET_TEXT'; payload: string; pushToUndo?: boolean }
+  | { type: 'APPLY_OPERATIONS'; payload: UpdateOperation[] }
+  | {
+      type: 'MOVE';
+      payload: {
+        dir: Direction;
+        visualLayout: ReturnType<typeof calculateVisualLayout>;
+      };
     }
-    if (newVisualScrollRow !== visualScrollRow) {
-      setVisualScrollRow(newVisualScrollRow);
+  | { type: 'DELETE' }
+  | { type: 'DELETE_WORD_LEFT' }
+  | { type: 'DELETE_WORD_RIGHT' }
+  | { type: 'KILL_LINE_RIGHT' }
+  | { type: 'KILL_LINE_LEFT' }
+  | { type: 'UNDO' }
+  | { type: 'REDO' }
+  | {
+      type: 'REPLACE_RANGE';
+      payload: {
+        startRow: number;
+        startCol: number;
+        endRow: number;
+        endCol: number;
+        text: string;
+      };
     }
-  }, [visualCursor, visualScrollRow, viewport]);
-
-  const pushUndo = useCallback(() => {
-    dbg('pushUndo', { cursor: [cursorRow, cursorCol], text: lines.join('\n') });
-    const snapshot = { lines: [...lines], cursorRow, cursorCol };
-    setUndoStack((prev) => {
-      const newStack = [...prev, snapshot];
-      if (newStack.length > historyLimit) {
-        newStack.shift();
-      }
-      return newStack;
-    });
-    setRedoStack([]);
-  }, [lines, cursorRow, cursorCol, historyLimit]);
-
-  const _restoreState = useCallback(
-    (state: UndoHistoryEntry | undefined): boolean => {
-      if (!state) return false;
-      setLines(state.lines);
-      setCursorRow(state.cursorRow);
-      setCursorCol(state.cursorCol);
-      return true;
-    },
-    [],
-  );
-
-  const text = lines.join('\n');
-
-  useEffect(() => {
-    if (onChange) {
-      onChange(text);
+  | { type: 'MOVE_TO_OFFSET'; payload: { text: string; offset: number } }
+  | { type: 'COPY' }
+  | { type: 'PASTE' }
+  | { type: 'START_SELECTION' };
+
+function textBufferReducer(
```

Comment:
suggestion(non-blocking): now with state transition captured in a pure function, we should be able to easily vibe-code up some clean unit tests for this. i'm not sure how we can guarantee this stays pure (ignoring `dbg`) with static analysis tools like eslint, so good tests are important here.